working-w-copilot/copilot-in-projects-new-code.md

```diff
@@ -0,0 +1,477 @@
+    }
+
+    with pytest.raises(ValueError):
+        Book.from_dict(book_data)
+```
+
+### !callout-warning
+
+## Creating `Book` instances in tests
+In our first test, `test_book_to_dict`, the `Book` class is instantiated from a dictionary with an explicit `id`, even though we discussed avoiding setting `id`s ourselves when creating the `from_dict` method. 
+
+<br />
+
+In this specific case, we're testing the `to_dict` method, which is isolated from our routes code or other database access. However we arrange the `Book` for testing this particular method is fine, but there are many instances where we need to use our fixtures to create Book instances that are tracked by SQLAlchemy and exist in our database for routes to retrieve. 
+
+<br />
+
+When we use Copilot to create tests, we should always evaluate if the data is being set up correctly in the `Arrange` step for all of the actions we will be taking.
+
+### !end-callout
+
+We have tests to ensure that for a given book model, the dictionary that `to_dict` creates contains the right data, and that a book created by `from_dict` stores the title and description from the input dictionary into the right properties. Our last test ensures that if the `"description"` key is missing from the input dictionary, we raise an error. But what if the `"title"` key is missing instead?
+
+To make our test suite as complete as possible, we can update the name of our last test to `test_book_from_dict_missing_description` and add one more test `test_book_from_dict_missing_title` to make sure that we have this edge case covered. We can write the additional test by hand, but we can also try highlighting the last test and asking Copilot to do that work for us!
+
+<br />
+
+<details>
+  <summary>
+    Try it out yourself, then expand this section to see how we asked Copilot to handle the test updates.
+  </summary>
+
+  **Updated prompt:**
+  > Please rename this test to reflect that it only tests what happens if the description field is missing. Next, add a new test for the from_dict function that tests the scenario where the title is missing but the description is present
+
+  **Updated test and newly added test:**
+  ```py
+  def test_book_from_dict_missing_description():
+      book_data = {
+          'title': 'The Hobbit'
+      }
+
+      with pytest.raises(ValueError):
+          Book.from_dict(book_data)
+
+  def test_book_from_dict_missing_title():
+      book_data = {
+          'description': 'A book about a hobbit'
+      }
+
+      with pytest.raises(ValueError):
+          Book.from_dict(book_data)
+  ```
+</details>
+
+### Creating `Author`, `Genre`, and `BookGenre` models
+
+We can ask Copilot to help us write something even if we don't have a template or example, but it's easier for Copilot to reason about what we want if we have samples to show. Here, we have a `Book` model that we can use as a pattern when asking Copilot to help us create the `Author`, `Genre`, and `BookGenre` models.
```

Comment:
- reworded to avoid claiming that Copilot "reasons" about anything
```suggestion
We can ask Copilot to help us write something even if we don't have a template or example, but Copilot tends to produce more relevant results if we have samples to show. Here, we have a `Book` model that we can use as a pattern when asking Copilot to help us create the `Author`, `Genre`, and `BookGenre` models.
```